### PR TITLE
feat: add program to calculate binomial coefficients

### PR DESCRIPTION
**Pull request by astronmax**
#### Description of Change

I have added a program to calculate binomial coefficients to the math module.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1448"></a>

### CODE REVIEW

File: math/binomial.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
+ * @brief Program to calculate binomial coefficients.
+ *
+ * @author astron
```
Comment (ayaankhan98) [inline math/binomial.cpp:5]: you can provide link to your github profile, if you wish.
Doxygen supports Markdown, you can use markdown like
`[Name](link to profile)`

**Review by ayaankhan98 — COMMENTED** [review]
rest of the code looks good to me.

**Comment by astronmax** [comment]
@ayaankhan98 Thanks, I made changes.

**Review by ayaankhan98 — APPROVED** [review]

File: math/binomial.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * @file
+ * @brief Program to calculate binomial coefficients.
+ *
+ * @author [astronmax](https://github.com/astronmax)
+ */
+
+#include <cassert>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
```
Comment (Panquesito7) [inline math/binomial.cpp:11]: Add a brief description of what the library adds (see the example below).
```c++
#include <iostream>  /// for IO operations
#include <assert>   /// for assert
```

File: math/binomial.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    return result;
+}
+
+/**
+ * TESTS
+ */
+
+void binomial_test() {
```
Comment (Panquesito7) [inline math/binomial.cpp:41]:
```suggestion
static void binomial_test() {
```

File: math/binomial.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        result *= n - k + i;
+        result /= i;
+    }
+
+    return result;
+}
+
+/**
+ * TESTS
+ */
```
Comment (Panquesito7) [inline math/binomial.cpp:39]:
```suggestion
/**
 * @brief Test implementations
 * @returns void
 */
```

File: math/binomial.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    assert(binomial(40, 20) == 137846528820);
+    assert(binomial(50, 25) == 126410606437752);
+    assert(binomial(60, 30) == 118264581564861424);
+    assert(binomial(62, 31) == 465428353255261088);
+
+    printf("[+] Binomial coefficients test completed\n");
+}
+
+/**
+ * Main function
```
Comment (Panquesito7) [inline math/binomial.cpp:58]:
```suggestion
 * @brief Main function
 * @returns 0 on exit
```

File: math/binomial.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    if (argc < 3) {
+        printf("Usage ./binomial {n} {k}\n");
+        return 0;
+    }
+
+    int32_t n = atoi(argv[1]);
+    int32_t k = atoi(argv[2]);
+
+    printf("%zd\n", binomial(n, k));
```
Comment (Panquesito7) [inline math/binomial.cpp:72]:
```suggestion
    printf("%zd\n", binomial(n, k));
    return 0;
```

File: math/binomial.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
+ * @brief Program to calculate binomial coefficients.
```
Comment (Panquesito7) [inline math/binomial.cpp:3]: Provide a Wikipedia link (if available/possible) in Markdown format.
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: math/binomial.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * @author [astronmax](https://github.com/astronmax)
+ */
+
+#include <cassert>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+
+/**
```
Comment (Panquesito7) [inline math/binomial.cpp:13]:
```suggestion
/**
 * @namespace math
 * @brief Mathematical algorithms
 */
namespace math {
/**
```

File: math/binomial.cpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        return 1;
+
+    size_t result = 1;
+    for (int32_t i = 1; i <= k; ++i) {
+        result *= n - k + i;
+        result /= i;
+    }
+
+    return result;
+}
```
Comment (Panquesito7) [inline math/binomial.cpp:35]:
```suggestion
}
}  // namespace math
```

File: math/binomial.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * TESTS
+ */
+
+void binomial_test() {
+    // test for n from 1 to 62
+    assert(binomial(1, 1) == 1);
```
Comment (Panquesito7) [inline math/binomial.cpp:43]:
```suggestion
    assert(math::binomial(1, 1) == 1);
```

File: math/binomial.cpp [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    assert(binomial(57, 57) == 1);
```
Comment (Panquesito7) [inline math/binomial.cpp:44]:
```suggestion
    assert(math::binomial(57, 57) == 1);
```

File: math/binomial.cpp [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    assert(binomial(6, 3) == 20);
+    assert(binomial(10, 5) == 252);
+    assert(binomial(20, 10) == 184756);
+    assert(binomial(30, 15) == 155117520);
+    assert(binomial(40, 20) == 137846528820);
+    assert(binomial(50, 25) == 126410606437752);
+    assert(binomial(60, 30) == 118264581564861424);
+    assert(binomial(62, 31) == 465428353255261088);
```
Comment (Panquesito7) [inline math/binomial.cpp:52]:
```suggestion
    assert(math::binomial(6, 3) == 20);
    assert(math::binomial(10, 5) == 252);
    assert(math::binomial(20, 10) == 184756);
    assert(math::binomial(30, 15) == 155117520);
    assert(math::binomial(40, 20) == 137846528820);
    assert(math::binomial(50, 25) == 126410606437752);
    assert(math::binomial(60, 30) == 118264581564861424);
    assert(math::binomial(62, 31) == 465428353255261088);
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Well done!
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/astronmax/C-Plus-Plus/actions

**Comment by astronmax** [comment]
@Panquesito7, thanks for review, i made changes.

File: math/binomial.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+#include <cassert>  /// for assert
+#include <cstdint>  /// for int32_t type
+#include <cstdio>   /// for printf
+#include <cstdlib>  /// for atoi
+
+/**
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
```
Comment (Panquesito7) [inline math/binomial.cpp:18]:
```suggestion
namespace math {
/**
 * @namespace binomial
 * @brief Functions for [Binomial
 * coefficients](https://en.wikipedia.org/wiki/Binomial_coefficient) implementation.
 */
namespace binomial {
```

File: math/binomial.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+    size_t result = 1;
+    for (int32_t i = 1; i <= k; ++i) {
+        result *= n - k + i;
+        result /= i;
+    }
+
+    return result;
+}
+}  // namespace math
```
Comment (Panquesito7) [inline math/binomial.cpp:42]:
```suggestion
}  // namespace binomial
}  // namespace math
```

File: math/binomial.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+
+static void binomial_test() {
+    // test for n from 1 to 62
+    assert(math::binomial(1, 1) == 1);
```
Comment (Panquesito7) [inline math/binomial.cpp:51]:
```suggestion
    assert(math::binomial::binomial(1, 1) == 1);
```

File: math/binomial.cpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(57, 57) == 1);
```
Comment (Panquesito7) [inline math/binomial.cpp:52]:
```suggestion
    assert(math::binomial::binomial(57, 57) == 1);
```

File: math/binomial.cpp [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(6, 3) == 20);
```
Comment (Panquesito7) [inline math/binomial.cpp:53]:
```suggestion
    assert(math::binomial::binomial(6, 3) == 20);
```

File: math/binomial.cpp [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(10, 5) == 252);
```
Comment (Panquesito7) [inline math/binomial.cpp:54]:
```suggestion
    assert(math::binomial::binomial(10, 5) == 252);
```

File: math/binomial.cpp [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(20, 10) == 184756);
```
Comment (Panquesito7) [inline math/binomial.cpp:55]:
```suggestion
    assert(math::binomial::binomial(20, 10) == 184756);
```

File: math/binomial.cpp [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(30, 15) == 155117520);
```
Comment (Panquesito7) [inline math/binomial.cpp:56]:
```suggestion
    assert(math::binomial::binomial(30, 15) == 155117520);
```

File: math/binomial.cpp [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(40, 20) == 137846528820);
+    assert(math::binomial(50, 25) == 126410606437752);
+    assert(math::binomial(60, 30) == 118264581564861424);
+    assert(math::binomial(62, 31) == 465428353255261088);
```
Comment (Panquesito7) [inline math/binomial.cpp:60]:
```suggestion
    assert(math::binomial::binomial(40, 20) == 137846528820);
    assert(math::binomial::binomial(50, 25) == 126410606437752);
    assert(math::binomial::binomial(60, 30) == 118264581564861424);
    assert(math::binomial::binomial(62, 31) == 465428353255261088);
```

File: math/binomial.cpp [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+    return result;
+}
+}  // namespace math
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+
+static void binomial_test() {
```
Comment (Panquesito7) [inline math/binomial.cpp:49]:
```suggestion
static void tests() {
```

File: math/binomial.cpp [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+    return result;
+}
+}  // namespace math
+
+/**
+ * @brief Test implementations
+ * @returns void
+ */
+
```
Comment (Panquesito7) [inline math/binomial.cpp:48]:
```suggestion
```

File: math/binomial.cpp [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+    assert(math::binomial(60, 30) == 118264581564861424);
+    assert(math::binomial(62, 31) == 465428353255261088);
+
+    printf("[+] Binomial coefficients test completed\n");
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
```
Comment (Panquesito7) [inline math/binomial.cpp:68]:
```suggestion
/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit
 */
```

File: math/binomial.cpp [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+int main(int argc, const char* argv[]) {
+    // run test
+    binomial_test();
```
Comment (Panquesito7) [inline math/binomial.cpp:71]:
```suggestion
    tests();  // run self-test implementations
```

File: math/binomial.cpp [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+
+    if (argc < 3) {
+        printf("Usage ./binomial {n} {k}\n");
+        return 0;
+    }
+
+    int32_t n = atoi(argv[1]);
+    int32_t k = atoi(argv[2]);
+
+    printf("%zd\n", math::binomial(n, k));
```
Comment (Panquesito7) [inline math/binomial.cpp:81]: As a C++ program, we use `std::cout` here, not C-style functions.
Please consider using C++-style functions as much as possible.
```suggestion
    printf("%zd\n", math::binomial::binomial(n, k));
```

File: math/binomial.cpp [hunk 27]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(int argc, const char* argv[]) {
+    // run test
+    binomial_test();
+
+    if (argc < 3) {
+        printf("Usage ./binomial {n} {k}\n");
```
Comment (Panquesito7) [inline math/binomial.cpp:74]: Same here, use `std::cout` instead.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Looks much better, excellent work! 👍
Almost there, just a few more suggestions.

**Comment by astronmax** [comment]
@Panquesito7, OK, Done. I decided to change the name of the file and the function, since, in my opinion, calling math::binomial::calculate() will look more logical than math::binomial::binomial().

**Review by Panquesito7 — APPROVED** [review]
Amazing work, I like how you've made the code. 😄 👍
Code and documentation are pretty good and refined; LGTM. 🎉

**Review by ayaankhan98 — APPROVED** [review]
@astronmax Thanks for your contribution

**Pull request merged**